File: bfs_dfs/team_proj.cpp
#include <iostream>
#include <queue>
using namespace std;

int student[100001];
int degree[100001];
bool visit[100001];

int BFS(int start) {
    visit[start] = true;
    queue<int> q;
    q.push(start);
    int cnt = 0;

    while(!q.empty()) {
        int now = q.front();
        visit[now] = true;
        q.pop();

        int parents = student[now];
        cnt++;
        
        degree[parents]--;

        if(degree[parents] == 0) {
            q.push(parents);
        }

    }

    return cnt;
}

int main(void) {
    
    int T;
    scanf("%d", &T);

    for(int t = 0; t < T; t++) {
        int N;
        scanf("%d", &N);

        for(int i = 1; i <= N; i++) {
            degree[i] = 0;
            visit[i] = false;
        }

        for(int i = 1; i <= N; i++) {
            scanf("%d", &student[i]);
            degree[student[i]]++; 
        }

        int ans = 0;
        for(int i = 1; i <= N; i++) {
            if(visit[i]) continue;

            if(degree[i] == 0) {
                ans += BFS(i);
            }
        }
        printf("%d\n", ans);
    }

    return 0;
}

File: bfs_dfs/team_proj.py
import sys
from collections import deque

input = sys.stdin.readline

T = int(input().rstrip())
def BFS(start, node, visit, degree) :
    d = deque()
    d.append(start)
    cnt = 0

    while d :
        now = d.popleft()
        visit[now] = True
        cnt += 1
        parents = node[now]
        degree[parents] -= 1

        if degree[parents] == 0 :
            d.append(parents)

    return cnt

for _ in range(T) :
    N = int(input().rstrip())
    people = list(map(int, input().rstrip().split()))
    people.insert(0,0)

    degree = [0] * (N + 1)
    ans = 0
    visit = [False] * (N + 1)

    for i in range(1, N + 1) :
        parents = people[i]
        degree[parents] += 1

    for i in range(1, N + 1) :
        if visit[i]: continue

        if degree[i] == 0:
            ans += BFS(i, people, visit, degree)

    print(ans)